libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ForEachElement.cs

```diff
@@ -226,12 +226,30 @@ private async Task<DialogTurnResult> RunItemsAsync(DialogContext dc, bool beginD
 
                 beginDialog = true;
                 UpdateActionScopeState(dc, new DialogState());
+
+                if (ShouldEndDialog(turnResult, out DialogTurnResult finalResult))
+                {
+                    return await dc.EndDialogAsync(result: finalResult, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
             }
 
             // End of list has been reached, or the list is null
             return await dc.EndDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
         }
 
+        private bool ShouldEndDialog(DialogTurnResult turnResult, out DialogTurnResult finalTurnResult)
+        {
+            finalTurnResult = turnResult;
```

Comment:
A comment to document why we decided to do this for future generations would help here, since we might forget this entire story in a few months.